Update iOS documentation to build 182 (#52)
Vualto/vudrm-public-docs: docs/source/integrations/mobile/ios-fairplay-sdk.md

 - Complete control over the entire AVPlayer lifecycle
 - Numerous optimisations to enable faster playback
 - Bitcode support

The SDK is fully supported in both Objective-C and Swift applications. 

> A multiple asset demo application written in Swift, and based on Apples FairPlay SDK example application, is available on request. Please contact [support@vualto.com](support@vualto.com) to request access.

Reviewer: ```suggestion
A multiple asset demo application written in Swift, and based on Apple's FairPlay SDK example application, is available on request. Please contact [support@vualto.com](support@vualto.com) to request access.
```

## Preparation

> The demo project is based on Apples FairPlay SDK example application. This is configured using the projects `Streams.plist` file. 

Reviewer: ```suggestion
The demo project is based on Apple's FairPlay SDK example application. This is configured using the projects `Streams.plist` file. 
```

To provide complete configuration for each `Stream` object in the `Streams.plist`you will require:

>  - An `skd://` URI entry in the `content_key_id_list`. The correct URI can be obtained by retrieving the manifest and parsing out the `URI` from the `EXT-X-SESSION-KEY` or `EXT-X-KEY`, this can be done in code or manually with curl in the Terminal app, eg enter: `curl https://eample.cloudfront.net/example-demo/examplecontent/examplecontent.ism/.m3u8`. 

Reviewer: ```suggestion
 - An `skd://` URI entry in the `content_key_id_list`. The correct URI can be obtained by retrieving the manifest and parsing out the `URI` from the `EXT-X-SESSION-KEY` or `EXT-X-KEY`, this can be done in code or manually with curl in the Terminal app, e.g. enter: `curl https://eample.cloudfront.net/example-demo/examplecontent/examplecontent.ism/.m3u8`. 
```

 - The `playlist_url` to correctly prepared content.

>  -  A `name` for the content (the correct content ID will be parsed out by VUDRMFairPlay SDK and will always be the last path component of the above `skd://` URI entry).

Reviewer: ```suggestion
 - A `name` for the content (the correct content ID will be parsed out by VUDRMFairPlay SDK and will always be the last path component of the above `skd://` URI entry).
```

 - A boolean value which indicates that the `Stream` object is protected with DRM. This setting would not be needed in scenarios where all content is protected, as all assets can be set as protected by default in the source code. 
 - A `vudrm-token` which must be correctly generated for the type of instance you wish to create. Example VUDRM token type templates available are:
	- Fairplay Rental `{"type": "r","duration_rental": 3600}`
		- Fairplay Rental tokens may only be used to stream online content to devices running iOS 9.x or higher.
	- Fairplay Persist `{"type": "p","duration_persist": 3600}`
 		- Fairplay Persist tokens may be used to stream online content and play offline (downloaded) content to devices running iOS 10.x or higher.

For further information about VUDRM please contact us, or refer to our documentation:
[https://docs.vualto.com/projects/vudrm/en/latest/VUDRM-token.html](https://docs.vualto.com/projects/vudrm/en/latest/VUDRM-token.html)

For added convenience, VUDRMFairPlay also bubbles up notifications of errors and progress during the licensing request.

Errors can be intercepted as follows:

>     @objc func handleVudrmDidError(_ notification: Notification)

Reviewer: Code should be in code blocks to make reading easier.

Author: In code blocks already, as discussed.

    {
        if let data = notification.userInfo as? [String: String]
        {
            for (function, error) in data
            {
                print("vudrmFairPlay - \(function) reported error \(error)!")
            }
        }
    }
    
And, progress can be monitored using the following:
   

>     @objc func handleVudrmProgressUpdate(_ notification: Notification)

Reviewer: Missing code block.

Author: In code blocks already, as discussed.

    {
        if let data = notification.userInfo as? [String: String]
        {
            for (function, message) in data
            {
                print("vudrmFairPlay - \(function) reported progress \(message).")
            }
        }
    }
        
## Example framework usage

> We strongly recommend referring to our example multiple asset demo application which uses a `ContentKeyManager`class to initialise an instance of VUDRMFairPlay for each stream, it then attaches the instance to an `AssetResourceLoaderDelegate`which is added to the `AssetResourceLoaderDelegateQueue`. Used with an `AssetPersistenceManager` class, the `ContentKeyManager` will also correctly handle persistence restoration with the `updateResourceLoaderDelegate` function.

Reviewer: ```suggestion
We strongly recommend referring to our example multiple asset demo application which uses a `ContentKeyManager`class to initialise an instance of VUDRMFairPlay for each stream, it then attaches the instance to an `AssetResourceLoaderDelegate` which is added to the `AssetResourceLoaderDelegateQueue`. Used with an `AssetPersistenceManager` class, the `ContentKeyManager` will also correctly handle persistence restoration with the `updateResourceLoaderDelegate` function.
```

	
The ContentKeyManager class should look like this:

> 	import AVFoundation

Reviewer: Missing code block.

Author: In code block already, as discussed.

 	import vudrmFairPlay
 	
 	class ContentKeyManager {
    
    /// MARK: Types.
    

>     /// The singleton for `ContentKeyManager`

Reviewer: I don't think inline code will work once this code is inside a code block.

Author: It works for me, as discussed.

     from `AVAssetResourceLoader`.
     */
    let assetResourceLoaderDelegate: vudrmFairPlay
    
    /// The DispatchQueue to use for delegate callbacks.
    let assetResourceLoaderDelegateQueue = DispatchQueue(label: "com.vualto.vurdrm.fairplay.AssetResourceLoaderDelegateQueue")
    
    var drm: vudrmFairPlay?
    
	/// MARK: Initialization.
    
    private init() {
        var token = ""
        var contentID = ""
        for stream in StreamListManager.shared.streams {
            if stream.vudrmToken != nil {
                token = stream.vudrmToken!
                contentID = stream.name
                let asset = AVURLAsset(url: URL(string: stream.playlistURL)!)
                self.drm = vudrmFairPlay(asset: asset, contentID: contentID, token: token)!
            }
        }
        assetResourceLoaderDelegate = self.drm!
    }
    
    func updateResourceLoaderDelegate(forAsset asset: AVURLAsset) {
        asset.resourceLoader.setDelegate(self.drm, queue: assetResourceLoaderDelegateQueue)
        }
    }
     
     
##### Online Playback / Download and Offline Playback:

> This will request a license for the content from the license server based on the type of token presented for each instance of VUDRM FairPlay. The tokens may be FairPlay Rental or FairPlay Persist. FairPlay Rental token license requests will be provided a streaming content key, which may be used for online streaming only. FairPlay Persist token license requests will provide a persistent content key, which may be used for both online streaming, and offline playback of downloaded content.

Reviewer: ```suggestion
This will request a license for the content from the license server based on the type of token presented for each instance of VUDRMFairPlay. The tokens may be FairPlay Rental or FairPlay Persist. FairPlay Rental token license requests will be provided a streaming content key, which may be used for online streaming only. FairPlay Persist token license requests will provide a persistent content key, which may be used for both online streaming, and offline playback of downloaded content.
```

When a download of an asset has been completed with a persist token policy, further playback requests will use the downloaded asset and associated content key, even when online.

There are significant limitations using AirPlay to stream any content that has been downloaded to the users device to an Apple TV. Please see the Limitations section for further information.

> To initialise your assets DRM instances, your application should create a DRM instance for each entry in the `Streams.plist` file (or comparable configuration) using a `ContentKeyManager`. How to do this is demonstrated in our demo application. The `ContentKeyManager` attaches instances of vudrmFairPlay to an `AssetResourceLoaderDelegate` for each asset using the asset URL, content ID and valid token. The content ID should be unique to each asset, as it is used to create a path to, and identify, offline assets and their associated content keys.

Reviewer: ```suggestion
To initialise your assets DRM instances, your application should create a DRM instance for each entry in the `Streams.plist` file (or comparable configuration) using a `ContentKeyManager`. How to do this is demonstrated in our demo application. The `ContentKeyManager` attaches instances of VUDRMFairPlay to an `AssetResourceLoaderDelegate` for each asset using the asset URL, content ID, and valid token. The content ID should be unique to each asset, as it is used to create a path to, and identify, offline assets and their associated content keys.
```

##### DEMO

> With the VUDRMFairPlay framework installed, the example demo application uses entries in the `Streams.plist` file to configure the DRM. You can edit the `Streams.plist` file values, and/or add your own valid values, ensuring the `content_key_id_list` entry is correct, and that your `token` is valid (see **Preparation** above).

Reviewer: If "Preparation" is a heading from above you should be able to link to so people can just click "Preparation" to jump to the relevant section.

Author: Full menu will be pushed after this has been merged, as discussed.

}
```

- In certain circumstances iOS 10 users activating Airplane mode can cause an offline asset to lose refernce to it's offline license.

- If you believe you have found any further issues, please contact us at support@vualto.com

## Release notes

### v1.0 (build 182) on 09/07/2020
- Added error and progress notifications.
- Improved error handling.

### v1.0 (build 161) on 10/06/2020
- SDK redeveloped to:

> 	-  resolve issue with persistence restoration after application cache clearance

Reviewer: ```suggestion
	- Resolve issue with persistence restoration after application cache clearance.
```

> 	-  enable demonstration of multiple asset DRM based on Apples HLS Catalog example application

Reviewer: ```suggestion
	- Enable demonstration of multiple asset DRM based on Apple's HLS Catalog example application.
```

> 	- improve flow and handling using Apples latest FairPlay SDK methods.

Reviewer: ```suggestion
	- Improve flow and handling using Apple's latest FairPlay SDK methods.
```
